src/log.c

```diff
@@ -167,10 +169,83 @@ static struct fi_ops_log ofi_default_log_ops = {
 	.log = ofi_log,
 };
 
+static int ofi_close_logging_fid(struct fid *fid)
+{
+	return 0;
+}
+
+static int ofi_bind_logging_fid(struct fid *fid, struct fid *bfid,
+				uint64_t flags);
+
+static struct fi_ops ofi_logging_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = ofi_close_logging_fid,
+	.bind = ofi_bind_logging_fid,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+	.tostr = fi_no_tostr,
+	.ops_set = fi_no_ops_set,
+};
+
 static struct fid_logging log_fid = {
+	.fid = {
+		.fclass = FI_CLASS_LOG,
+		.ops = &ofi_logging_ops,
+	},
 	.ops = &ofi_default_log_ops,
 };
 
+static int ofi_close_import(struct fid *fid)
+{
+	/* Reset logging ops to default */
+	pthread_mutex_lock(&common_locks.ini_lock);
+	log_fid.ops = &ofi_default_log_ops;
+	log_fid.fid.context = NULL;
+	pthread_mutex_unlock(&common_locks.ini_lock);
+	return 0;
+}
+
+static struct fi_ops impfid_ops = {
+	.size = sizeof(struct fi_ops),
+	.close = ofi_close_import,
+	.bind = fi_no_bind,
+	.control = fi_no_control,
+	.ops_open = fi_no_ops_open,
+	.tostr = fi_no_tostr,
+	.ops_set = fi_no_ops_set,
+};
+
+static int ofi_logging_import(struct fid *fid)
+{
+	struct fid_logging *impfid;
+
+	if (fid->fclass != FI_CLASS_LOG)
+		return -FI_EINVAL;
+
+	impfid = container_of(fid, struct fid_logging, fid);
+	if (impfid->ops->size < sizeof(struct fi_ops_log))
+		return -FI_EINVAL;
+
+	if (impfid->ops->enabled)
+		log_fid.ops->enabled = impfid->ops->enabled;
+	if (impfid->ops->ready)
+		log_fid.ops->ready = impfid->ops->ready;
+	if (impfid->ops->log)
+		log_fid.ops->log = impfid->ops->log;
```

Comment:
These assignments end up modifying the pointers in ofi_default_log_ops.  I would remove the 'default' from the name since the functions may not be the default.

Reply:
Thanks for catching that. I reworked this part.